core/trino-spi/src/test/java/io/trino/spi/type/TestInt128Math.java

```diff
@@ -0,0 +1,818 @@
+        BigInteger rescaledDividend = dividendBigInteger.multiply(bigIntegerTenToNth(dividendRescaleFactor));
+        BigInteger rescaledDivisor = divisorBigInteger.multiply(bigIntegerTenToNth(divisorRescaleFactor));
+
+        BigInteger expectedQuotient = new BigDecimal(rescaledDividend).divide(new BigDecimal(rescaledDivisor), RoundingMode.HALF_UP).toBigIntegerExact();
+        BigInteger expectedRemainder = rescaledDividend.remainder(rescaledDivisor);
+
+        boolean overflowIsExpected = expectedQuotient.abs().compareTo(Int128.MAX_VALUE.toBigInteger()) >= 0 || expectedRemainder.abs().compareTo(Int128.MAX_VALUE.toBigInteger()) >= 0;
+
+        Int128 quotient;
+        Int128 remainder;
+        try {
+            quotient = divideRoundUp(
+                    dividend.getHigh(),
+                    dividend.getLow(),
+                    dividendRescaleFactor,
+                    divisor.getHigh(),
+                    divisor.getLow(),
+                    divisorRescaleFactor);
+
+            remainder = remainder(
+                    dividend.getHigh(),
+                    dividend.getLow(),
+                    dividendRescaleFactor,
+                    divisor.getHigh(),
+                    divisor.getLow(),
+                    divisorRescaleFactor);
+
+            if (overflowIsExpected) {
+                fail("overflow is expected");
+            }
+
+            BigInteger actualQuotient = quotient.toBigInteger();
+            BigInteger actualRemainder = remainder.toBigInteger();
+
+            if (expectedQuotient.equals(actualQuotient) && expectedRemainder.equals(actualRemainder)) {
+                return;
+            }
+
+            fail(format("%s / %s ([%s * 2^%d] / [%s * 2^%d]) Expected: %s(%s). Actual: %s(%s)",
+                    rescaledDividend, rescaledDivisor,
+                    dividendBigInteger, dividendRescaleFactor,
+                    divisorBigInteger, divisorRescaleFactor,
+                    expectedQuotient, expectedRemainder,
+                    actualQuotient, actualRemainder));
+        }
+        catch (ArithmeticException e) {
+            if (!overflowIsExpected) {
+                fail("overflow wasn't expected");
+            }
+        }
+    }
+
+    private static boolean isShort(BigInteger value)
+    {
+        return value.abs().shiftRight(63).equals(BigInteger.ZERO);
+    }
+
+    private static void assertMultiply(BigInteger a, long b, BigInteger result)
+    {
+        assertMultiply(a, BigInteger.valueOf(b), result);
+    }
+
+    private static void assertMultiply(BigInteger a, long b, long result)
+    {
+        assertMultiply(a, BigInteger.valueOf(b), BigInteger.valueOf(result));
+    }
+
+    private static void assertMultiply(long a, long b, BigInteger result)
+    {
+        assertMultiply(BigInteger.valueOf(a), b, result);
+    }
+
+    private static void assertMultiply(long a, long b, long result)
+    {
+        assertMultiply(a, b, BigInteger.valueOf(result));
```

Comment:
If the implementation is supposed to match `BigInteger` you could use that to create the result value.

Reply:
It's preexistent. I just renamed the class.